physicalplan/aggregate/scalar_table.go

```diff
@@ -118,81 +112,100 @@ type accumulator struct {
 	Reset     func()
 }
 
-func newAccumulator(expr parser.ItemType) (*accumulator, error) {
-	hasValue := false
+func newAccumulator(expr parser.ItemType) (newAccumulatorFunc, error) {
 	t := parser.ItemTypeStr[expr]
 	switch t {
 	case "sum":
-		var value float64
-		return &accumulator{
-			AddFunc: func(v float64) {
-				hasValue = true
-				value += v
-			},
-			ValueFunc: func() float64 { return value },
-			HasValue:  func() bool { return hasValue },
-			Reset: func() {
-				hasValue = false
-				value = 0
-			},
+		return func() *accumulator {
+			var value float64
+			var hasValue bool
+
+			return &accumulator{
+				AddFunc: func(v float64) {
+					hasValue = true
+					value += v
+				},
+				ValueFunc: func() float64 { return value },
+				HasValue:  func() bool { return hasValue },
+				Reset: func() {
+					hasValue = false
+					value = 0
+				},
+			}
 		}, nil
 	case "max":
-		var value float64
-		return &accumulator{
-			AddFunc: func(v float64) {
-				hasValue = true
-				value = math.Max(value, v)
-			},
-			ValueFunc: func() float64 { return value },
-			HasValue:  func() bool { return hasValue },
-			Reset: func() {
-				hasValue = false
-				value = 0
-			},
+		return func() *accumulator {
+			var value float64
+			var hasValue bool
+
+			return &accumulator{
+				AddFunc: func(v float64) {
+					hasValue = true
+					value = math.Max(value, v)
```

Comment:
We need to check `hasValue` in case v < 0 ? Or initialize value as the min float64

Reply:
I think you're right. Will create an issue for this 👍

Reply:
Created https://github.com/thanos-community/promql-engine/issues/37